x-pack/plugins/lens/public/app_plugin/app.test.tsx

```diff
@@ -1001,4 +952,135 @@ describe('Lens App', () => {
 
     expect(args.core.notifications.toasts.addDanger).toHaveBeenCalled();
   });
+
+  describe('showing a confirm message when leaving', () => {
+    let defaultArgs: ReturnType<typeof makeDefaultArgs>;
+    let defaultLeave: jest.Mock;
+    let confirmLeave: jest.Mock;
+
+    beforeEach(() => {
+      defaultArgs = makeDefaultArgs();
+      defaultLeave = jest.fn();
+      confirmLeave = jest.fn();
+      (defaultArgs.docStorage.load as jest.Mock).mockResolvedValue({
+        id: '1234',
+        title: 'My cool doc',
+        expression: 'valid expression',
+        state: {
+          query: 'kuery',
+          datasourceMetaData: { filterableIndexPatterns: [{ id: '1', title: 'saved' }] },
+        },
+      } as jest.ResolvedValue<Document>);
+    });
+
+    it('should not show a confirm message if there is no expression to save', () => {
```

Comment:
I'm not sure whether this is expected, because the expression is also empty if the current configuration is invalid. E.g.:
* Create count by date histogram split by some terms
* Save
* Remove the split
* It won't allow you to navigate away, because you have unsaved (valid) changes

* Create count by date histogram split by some terms
* Save
* Remove the count from y axis
* It will allow you to navigate away, because your changes are currently invalid

I think this PR is mostly sound, so I wouldn't block on this, but I wanted to mention it. Maybe there is a better way to check whether changes have happened (deep equal of the state tree?). It doesn't have to be fast because it's only happening when navigating away.

Reply:
Okay, I changed the logic here. I am still looking at the expression primarily, but this now handles the case where the user enters the invalid state.